### PR TITLE
Change how providers are selected with pkey operations

### PR DESCRIPTION
**Pull request by simo5**
This PR changes the selection logic for finding the correct provider when initializing operations that use pkeys.

The TL;DR is: OpenSSL should prefer to use the same provider that was used to fetch the keys when selecting an operation handler. And fallback to searching more generic providers only if that provider cannot handle the requested operation.

I ran through this issues as I am writing this provider:
https://github.com/latchset/pkcs11-provider

A peculiar thing with pkcs11 providers is that it is always very obvious that keys come from a pkcs11 token because you need to use a pkcs11: URI to source them, and that selects automatically the pkcs11 provider store functions which will give back a pkcs11 provider specific key.
This means that by just passing a URI into openssl commands without specifying a provider or propq strings you end up using the pkcs11-provider and obtain a valid key handler.

However after obtaining a key I found out, earlier in development, that all operations were failing with odd errors due to key export issues (private keys cannot be exported from a pkcs11 token).
Not knowing any better I initially worked around this problem by returning keys that had a special data_type (for example "PKCS11-RSA" instead of "RSA") and exporting from keymgmt and other operations under both the standard and the special name ...
I now have a PR to rectify the situation depending on this PR, which I have tested with:
https://github.com/latchset/pkcs11-provider/pull/67

Why change selection ?

In general I think it makes more sense to stick to the provider that gave you keys by default, it just seem to make more sense. However there are actual practical benefits in terms of applications.

Having to always pass a "provider=whatever" property to *all* EVP_PKEY_something() calls is somewhat burdensome for existing applications ported over from the OpenSSL 1.1 days, it would definitely be simpler to just change those calls that fetch keys, and then assume that if you got a key from provider X, whenever you do an operation on that key, that provider X is being used by default.

For example, an application wanting to use FIPS where possible, but ok to make exception where not (as stop gap to compliance) would want to avoid setting the provider=fips selection in the configuration file and rather do it in code where needed.

Without the current PR it would have the daunting tasks to change all invocations and carefully manually set the propq everywhere. Because at every turn the current code will always select the 'default' provider in preference, regardless of which provider's keymgmt functions were used to source or generate the keys.

Another case is where an application load keys directly but then uses libraries that themselves use openssl. With this PR the application can make sure the key is loaded with a provider of choice (for example a provider that uses a hw accelerator), and then not need to change the 3rd party library to have it use the "preferred" provider by default. This is especially important when the acceleration engine offers only part of the functionality. In that case the desired behavior is to use the acceleration provider where possible, but fallback to the 'default' provider otherwise.

Finally I could not find a reason to not select the "closest" provider first, and only later fallback to a generic one, that doesn't mean a reason does not exist, I just couldn't find one.

### CODE REVIEW

**Review by paulidale — APPROVED** [review]
LGTM

**Review by slontis — COMMENTED** [review]
Setting up a test that uses a  dummy provider to test this might be a good idea.

**Comment by levitte** [comment]
I'd like to see a written entry in CHANGES.md.

Unfortunately, that's forgotten, much too often...

**Comment by romen** [comment]
@paulidale Approved as a bugfix or a feature? Which branches are targeted?

**Comment by paulidale** [comment]
I hadn't decided if I thought was a bug fix or not.
It will change behaviour but not in anyway we guarantee.

**Comment by t8m** [comment]
The reason we did it this way was that the preference set by the property query should be followed _if it is actually possible_. I.e., the fallback to the fetch from the provider should happen only when the export is not possible. Otherwise, with this PR, if you for example specify "?provider=fips" in the query, and you fetch an exportable key from some other provider it will use the operation from the other provider although the fips provider is preferred.

My question would be, why the existing order does not work for you? If the key is not exportable, the fallback fetch should work fine.

**Comment by t8m** [comment]
If we decide to change the order, IMO it is not a bug fix but a feature.

**Comment by romen** [comment]
> This points me towards bugfix and 3.0:
> 
> > However after obtaining a key I found out, earlier in development, that all operations were failing with odd errors due to key export issues.

@simo5 could you assist us in creating a simple reproducer of the issues you experienced?

**Comment by beldmit** [comment]
@t8m do we consider that export is smth in OpenSSL internal format suitable for decoder? 

E.g. PKCS#11 provides opaque handles with URI as a proper identifier. Do we consider it as export?

**Comment by t8m** [comment]
> E.g. PKCS#11 provides opaque handles with URI as a proper identifier. Do we consider it as export?

Nope. The keymgmt export function should export something that other providers supporting the same algorithm should understand.

**Comment by romen** [comment]
> The export function in a provider should fail when it is unable to export all the parts of the key that the selection asks for and the key has. It should never successfully export just a public key part if `selection == OSSL_KEYMGMT_SELECT_KEYPAIR`, and the private key part is there but is not exportable.

@t8m are these requirements properly documented? IIRC the `selection` logic is not handled in the same way in all the functions that support it (i.e., sometimes it's akin to an AND rather than a OR, sometimes we are strict in returning exactly what was requested or an error if anything is missing, others we return what we can to satisfy the request at least partially, etc.).

Looking at https://www.openssl.org/docs/man3.0/man7/provider-keymgmt.html I don't see enough information about your remark.

**Comment by t8m** [comment]
If you want to encode a blob that can be later used with your decoder to decode and identify an internal key with it, you need to use encoder from the same provider.

**Comment by t8m** [comment]
@romen I agree that this particular constraint is probably not documented.

**Comment by romen** [comment]
> @romen I agree that this particular constraint is probably not documented.

@t8m I actually believe that is is worse than that at the moment, and that some of the current info about the _selector bits_, might be even read as contradicting on the restriction you described for `OSSL_FUNC_keymgmt_export()`:

> **It's left to the provider implementation to decide what is reasonable to do with regards to received selector bits and how to do it**. Among others, an implementation of OSSL_FUNC_keymgmt_match() might opt to not compare the private half if it has compared the public half, since a match of one half implies a match of the other half.

**Comment by romen** [comment]
> The reason we did it this way was that the preference set by the property query should be followed _if it is actually possible_. I.e., the fallback to the fetch from the provider should happen only when the export is not possible. Otherwise, with this PR, if you for example specify "?provider=fips" in the query, and you fetch an exportable key from some other provider it will use the operation from the other provider although the fips provider is preferred.
> 
> My question would be, why the existing order does not work for you? If the key is not exportable, the fallback fetch should work fine.

This makes me think maybe this PR should not be accepted, and changing the order can cause loss of functionality in some specific cases depending on `propquery`.

From the discussion it seems that the best way to avoid the errors @simo5 was seeing would rather be to avoid implementing  `OSSL_FUNC_keymgmt_export()` in the `pkcs11` provider, or change its current implementation to return an error if any of the  bits in the `selection` argument cannot be exported in a way understood by core/other providers.

**Comment by t8m** [comment]
> The emphasis in the excerpt above is mine: I could read the first sentence as mandating that the logic on how strict to be on interpreting the selection bits is very much left to provider implementations, given it is presented in the description of the generic concepts about key management and key objects, before specific functions, and that the specifics of the section on import/export functions does not mandate any ulterior requirement or guideline.

Sure, this is a general recommendation which gives full power to implementors of providers to make it not work properly with the current libcrypto implementation. :grin:

**Comment by beldmit** [comment]
> Having to always pass a "provider=whatever" property to _all_ EVP_PKEY_something() calls is somewhat burdensome for existing applications ported over from the OpenSSL 1.1 days, it would definitely be simpler to just change those calls that fetch keys, and then assume that if you got a key from provider X, whenever you do an operation on that key, that provider X is being used by default.

I consider this a pretty good rationale.

**Comment by beldmit** [comment]
> > E.g. PKCS#11 provides opaque handles with URI as a proper identifier. Do we consider it as export?
> 
> Nope. The keymgmt export function should export something that other providers supporting the same algorithm should understand.

So public keys is exportable and private is not in this sense

**Comment by t8m** [comment]
> So public keys is exportable and private is not in this sense

Yes, but you should not treat the private key as a public key with the export function when the selection includes OSSL_KEYMGMT_SELECT_PRIVATE_KEY.

**Comment by t8m** [comment]
> I consider this a pretty good rationale.

Yes, but it is changing a behavior and definitely not a bug fix.

If we designed it like this from scratch it would be a different question.

**Comment by simo5** [comment]
> From the discussion it seems that the best way to avoid the errors @simo5 was seeing would rather be to avoid implementing `OSSL_FUNC_keymgmt_export()` in the `pkcs11` provider, or change its current implementation to return an error if any of the bits in the `selection` argument cannot be exported in a way understood by core/other providers.

My provider already returns 0 (unconditioanlly for EC keys for example) except for RSA public keys which I *can* export. 

It does not work.

**Comment by simo5** [comment]
> Yes, but you should not treat the private key as a public key with the export function when the selection includes OSSL_KEYMGMT_SELECT_PRIVATE_KEY.

And my provider is not, keep in mind in most of my test cases OpenSSL has a specific key type preference so it never passes in both in the selection. Nowehere the code seem to work as you expect, returning an export error simply causes the whole thing to fail.

**Comment by romen** [comment]
> > From the discussion it seems that the best way to avoid the errors @simo5 was seeing would rather be to avoid implementing `OSSL_FUNC_keymgmt_export()` in the `pkcs11` provider, or change its current implementation to return an error if any of the bits in the `selection` argument cannot be exported in a way understood by core/other providers.
> 
> My provider already returns 0 (unconditioanlly for EC keys for example) except for RSA public keys which I _can_ export.
> 
> It does not work.

https://github.com/latchset/pkcs11-provider/blob/main/src/keymgmt.c#L545-L547

this does not return an error if `selection == OSSL_KEYMGMT_SELECT_KEYPAIR`

**Comment by simo5** [comment]
> > > From the discussion it seems that the best way to avoid the errors @simo5 was seeing would rather be to avoid implementing `OSSL_FUNC_keymgmt_export()` in the `pkcs11` provider, or change its current implementation to return an error if any of the bits in the `selection` argument cannot be exported in a way understood by core/other providers.
> > 
> > 
> > My provider already returns 0 (unconditioanlly for EC keys for example) except for RSA public keys which I _can_ export.
> > It does not work.
> 
> https://github.com/latchset/pkcs11-provider/blob/7782f12b90e84bb4ff3c2699b4f41f5e7c75d107/src/keymgmt.c#L545-L547
> 
> this does not return an error if `selection == OSSL_KEYMGMT_SELECT_KEYPAIR`

This is true for RSA, but if you look at EC I always return an error.
I will however make sure to return an error unless exclusively a public key selection is provided and see if there is a fallback for RSA and report.

**Comment by simo5** [comment]
Just a heads up, this stuff is broken until OpenSSL 3.0.5 where this commit af40029ea9bbc7289b05a8af84911b10cc442576 was added ...
